### ISSUE TITLE
Client interfaces should not expose internal rest clients

### ISSUE THREAD

**Issue by liggitt** [kubernetes/kubernetes#26130]
https://github.com/kubernetes/kubernetes/pull/24220 added getters to expose internal RESTClient objects in Client interfaces. That tightly couples those interfaces to a particular implementation, which doesn't make sense (as evidenced by the fake clients returning nil, and anything calling those getters having to handle nils)

cc @krousey @smarterclayton @gmarek

**Comment by smarterclayton** [comment]
Yeah, originally that was abstracted intentionally so they were testable.
If we had a factory approach for controller clients that would alleviate
some of this (by letting us pass the rate limiter in).

We're going to have to open up construction of the client a bit more anyway
(since Config was not intended to be a straightjacket).

**Comment by caesarxuchao** [comment]
subscribe.

**Comment by gmarek** [comment]
cc @lavalamp

**Comment by sftim** [comment]
We still need to do this cleanup, right?

**Comment by liggitt** [comment]
I _really_ want to, almost every use of this method is either wrong or problematic in tests, but am not sure we can drop it non-disruptively

A good first step would be to eradicate use in-tree and mark the method deprecated / document it as problematic

**Comment by pranav-pandey0804** [comment]
Hello @liggitt,
I'd like to assist with the initial steps you mentioned regarding the `RESTClient()` method. Should I start by auditing in-tree uses to assess its impact?

**Comment by pranav-pandey0804** [comment]
Following up, if there is a list of components that currently depend on this method, could you share it?

**Comment by pranav-pandey0804** [comment]
@liggitt  Once we have a clear understanding of the usage, I'm also ready to help document the deprecation notice and outline the reasons it's considered problematic.

**Comment by pranav-pandey0804** [comment]
also in anticipation of potential concerns about the deprecation, do we have any strategies or alternatives in mind to mitigate disruptions for users currently reliant on the `RESTClient()` method?
